### PR TITLE
Feat/padding

### PR DESCRIPTION
**Pull request by Kubuxu**

### CODE REVIEW

File: parsing/padding.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
+package parsing
+
+import (
+	"errors"
+	"github.com/filecoin-project/go-data-segment/types"
+)
+
+// Pad pads a general byte array in to FP32 chunks of bytes where the topmost bits of the most significant byte are 0
+func Pad(unpaddedData *[]byte) ([]types.FP32, error) {
```
Comment (Kubuxu) [inline parsing/padding.go]: You don't need to take the pointer here. The slice is a tuple of (*array, len, capacity) and is usually passed by value.

File: parsing/padding.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
+package parsing
+
+import (
+	"errors"
+	"github.com/filecoin-project/go-data-segment/types"
+)
+
+// Pad pads a general byte array in to FP32 chunks of bytes where the topmost bits of the most significant byte are 0
```
Comment (Kubuxu) [inline parsing/padding.go]: Nit: it isn't FP32 but Fr32, it gets its name from the `r` of BLS-12-381: https://github.com/filecoin-project/paired/tree/master/src/bls12_381#bls12-381-instantiation.
It is a padding scheme compatible with Field over `r`.

File: parsing/padding.go [hunk 3]
```diff
@@ -2,103 +2,103 @@ package parsing
 
 import (
 	"errors"
-	"github.com/filecoin-project/go-data-segment/types"
+	"github.com/filecoin-project/go-data-segment/fr32"
```
Comment (Kubuxu) [inline parsing/padding.go:5]: I think `types` previously was fine, we can have more types there and now we have a repetition of the `fr32.Fr32` which is a bit ugly.

Comment (Kubuxu) [reply parsing/padding.go:5]: Ahh, I see why you did it, so the Bits* are nicer.

Comment (Kubuxu) [reply parsing/padding.go:5]: This works as well, I would say.

File: parsing/padding.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		unpaddedChunk := getChunk(bitIdx, unpaddedData)
+		paddedData[i] = fr32.Fr32{Data: shiftChunk(bitIdx, unpaddedChunk)}
+		// Update bitIdx to the byte we need to start at which is 254 in
+		bitIdx += fr32.BitsNeeded
+	}
+	return paddedData, nil
+}
+
+// Return a slice containing the next segment of unpadded data (without copying data), it will be a slice of either 32 or 33 bytes
+func getChunk(bitIdx int, unpaddedData *[]byte) []byte {
```
Comment (Kubuxu) [inline parsing/padding.go]: The pointer to slice snuck through.

File: parsing/padding.go (hunk 2 above)
Comment (jot2re) [reply parsing/padding.go]: Fixed in commit cea535e1000d13d5d4ffaf21eb601b3731553bee

File: parsing/padding.go (hunk 4 above)
Comment (jot2re) [reply parsing/padding.go]: Thanks! Fixed in d2c13c6e41f87ffbf9e808d71a6271df03a274cc

File: parsing/padding.go (hunk 1 above)
Comment (jot2re) [reply parsing/padding.go]: Thanks for the clarification, that is a very nice detail to know about arrays in Go!

Comment (jot2re) [reply parsing/padding.go]: Fixed in 9872fe3b0c52186e745d15359c1a8ab15c0fd486

**Comment by jot2re** [comment]
@Kubuxu I implemented the fixes suggested. I can unfortunately not set you explicitly as the reviewer as you created the PR

**Pull request merged**